### PR TITLE
Weather Click

### PR DESCRIPTION
**Pull request by rafcio2k**

### CODE REVIEW

**Comment by francois-berder** [comment]
First of all, you need to solve the conflict in CMakeLists.txt

File: src/click/weather.c [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+ WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+ USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ ************************************************************************************************************************/
+
+#include <letmecreate/letmecreate.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <time.h>
+
+struct bme280_t bme280;
```
Comment (francois-berder) [inline src/click/weather.c]: should be static

File: src/click/weather.c [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+static uint8_t readResult = 0;
+
+s8 weather_click_i2c_bus_read(u8 dev_addr, u8 reg_addr, u8 *reg_data, u8 cnt) {
```
Comment (francois-berder) [inline src/click/weather.c]: In src/click/weather.c and include/letmecreate/click/weather.h, do not use s8, s32, u32. Instead, use uint8_t, int8_t, uint32_t or int32_t.

File: src/click/weather.c [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+    req.tv_sec = msek / 1000;
+    msek -= req.tv_sec * 1000;
+    req.tv_nsec = msek * 1000000;
+
+    while (nanosleep(&req, &rem)) {
+        req = rem;
+    }
+}
+
+uint8_t weather_click_read_measurements(double* temperature, double* pressure, double* humidity) {
```
Comment (francois-berder) [inline src/click/weather.c]: { should be on next line for each function implementation

File: src/click/weather.c [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+
+uint8_t weather_click_enable() {
+    readResult = 0;
+
+    bme280.bus_write = weather_click_i2c_bus_write;
+    bme280.bus_read = weather_click_i2c_bus_read;
+    bme280.dev_addr = BME280_I2C_ADDRESS1;
+    bme280.delay_msec = weather_click_delay_msek;
+
+    if (bme280_init(&bme280) != 0) {
```
Comment (francois-berder) [inline src/click/weather.c]: Use the following pattern instead:

```c
if (bme280_init(&bme280) != 0
||  bme280_set_power_mode(BME280_SLEEP_MODE) != 0
 ...
    return -1
```

This is much more compact.

File: include/letmecreate/click/weather.h [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+/**
+ * @brief Reads current values of temperature, pressure, humidity from sensor.
+ *
+ * @param[out] temperature measured by sensor, or NULL if not interested in this value.
+ * @param[out] pressure measured by sensor, or NULL if not interested in this value.
+ * @param[out] humidity measured by sensor, or NULL if not interested in this value.
+ * @return 0 if success otherwise -1
+ */
+uint8_t weather_click_read_measurements(double* temperature, double* pressure, double* humidity);
```
Comment (francois-berder) [inline include/letmecreate/click/weather.h]: return int instead of uint8_t.

File: include/letmecreate/click/weather.h [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+/**
+ * @brief Setups sensor configuration as recommended for indoor measurement in Bosh BME 280 data sheet.
+ * @return 0 if success otherwise -1
+ */
+uint8_t weather_click_enable();
```
Comment (francois-berder) [inline include/letmecreate/click/weather.h]: Same, return int instead of uint8_t

File: include/letmecreate/click/weather.h [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+
```
Comment (francois-berder) [inline include/letmecreate/click/weather.h]: No weather_click_disable ?

File: src/click/weather.c [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+        if (iError < 0) {
+            readResult = -1;
+            break;
+        }
+    }
+
+    return (s8) iError;
+}
+
+void weather_click_delay_msek(u32 msek) {
```
Comment (francois-berder) [inline src/click/weather.c]: This function should be static

File: src/click/weather.c [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+        if (iError < 0) {
+            readResult = -1;
+            break;
+        }
+    }
+
+    return (s8) iError;
+}
+
+s8 weather_click_i2c_bus_write(u8 dev_addr, u8 reg_addr, u8 *reg_data, u8 cnt) {
```
Comment (francois-berder) [inline src/click/weather.c]: Function should be static

File: src/click/weather.c [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+uint8_t weather_click_read_measurements(double* temperature, double* pressure, double* humidity) {
+
+    readResult = 0;
+    s32 uncompTemp = BME280_INIT_VALUE;
+    s32 uncompPress = BME280_INIT_VALUE;
+    s32 uncompHumidity = BME280_INIT_VALUE;
+
+    uint8_t operationResult = bme280_read_uncomp_pressure_temperature_humidity(&uncompPress, &uncompTemp, &uncompHumidity);
+
+    if (temperature != NULL) {
```
Comment (francois-berder) [inline src/click/weather.c]: When the body of the condition is only one line, do not use curly braces

File: src/click/weather.c [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+        }
+    }
+
+    return (s8) iError;
+}
+
+void weather_click_delay_msek(u32 msek) {
+    struct timespec req, rem;
+
+    req.tv_sec = msek / 1000;
```
Comment (francois-berder) [inline src/click/weather.c]: rename msek to ms

File: src/click/weather.c (hunk 1 above)
Comment (rafcio2k) [inline src/click/weather.c]: done

File: src/click/weather.c (hunk 9 above)
Comment (rafcio2k) [inline src/click/weather.c]: done

File: src/click/weather.c (hunk 8 above)
Comment (rafcio2k) [inline src/click/weather.c]: done

File: src/click/weather.c (hunk 11 above)
Comment (rafcio2k) [inline src/click/weather.c]: done

File: src/click/weather.c (hunk 3 above)
Comment (rafcio2k) [inline src/click/weather.c]: done

File: src/click/weather.c (hunk 10 above)
Comment (rafcio2k) [inline src/click/weather.c]: done

File: include/letmecreate/click/weather.h (hunk 5 above)
Comment (rafcio2k) [inline include/letmecreate/click/weather.h]: done

File: include/letmecreate/click/weather.h (hunk 6 above)
Comment (rafcio2k) [inline include/letmecreate/click/weather.h]: done

File: include/letmecreate/click/weather.h (hunk 7 above)
Comment (rafcio2k) [inline include/letmecreate/click/weather.h]: Added weather_click_disable function.

File: src/click/weather.c (hunk 2 above)
Comment (rafcio2k) [inline src/click/weather.c]: change back to u8 to stay consistent with bme280 declaration:
`#define BME280_RD_FUNC_PTR\
		s8 (*bus_read)(u8, u8,\
		u8 *, u8)`

**Pull request merged**